File: 23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);
        for(auto* l : lists) {
            if(l != nullptr) {
                pq.push(l);
            }
        }

        ListNode dummy, *tail = &dummy;
        while(!pq.empty()) {
            auto* curr = pq.top();
            pq.pop();

            tail->next = curr;
            tail = curr;
            if(curr->next) {
                pq.push(curr->next);
            }
        }
        return dummy.next;
    }
};
// @lc code=end

File: 23.merge-k-sorted-lists.py
#
# @lc app=leetcode id=23 lang=python3
#
# [23] Merge k Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        for i, l in enumerate(lists):
            if l:
                heapq.heappush(min_heap, (l.val, i, l))
        
        dummy = ListNode()
        curr = dummy
        while min_heap:
            _, level, node = heapq.heappop(min_heap)
            curr.next = node
            if node.next:
                heapq.heappush(min_heap, (node.next.val, level, node.next))
            curr = curr.next
        
        return dummy.next
# @lc code=end

